libkbfs/rekey_fsm.go

```diff
@@ -0,0 +1,488 @@
+			})
+		fsm.log.CDebugf(ctx, "Rekey finished with res=%#+v, error=%v", res, err)
+		fsm.Event(newRekeyFinishedEvent(res, err))
+		if cancel != nil {
+			cancel()
+		}
+	}()
+	return &rekeyStateStarted{
+		fsm:  fsm,
+		task: task,
+	}
+}
+
+func (r *rekeyStateStarted) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyFinishedEvent:
+		ttl := r.task.ttl - 1
+		r.fsm.log.CDebugf(r.task.ctx,
+			"Rekey Finished. ttl: %d -> %d", r.task.ttl, ttl)
+
+		if ttl <= 0 {
+			r.fsm.log.CDebugf(r.task.ctx,
+				"Not scheduling new rekey because TTL expired.")
+			return newRekeyStateIdle(r.fsm)
+		}
+
+		switch event.finished.err {
+		case nil:
+		default:
+			r.fsm.log.CDebugf(r.task.ctx, "Rekey errored. Scheduling new rekey.")
+			return newRekeyStateScheduled(r.fsm, 0, rekeyTask{
+				timeout:     r.task.timeout,
+				promptPaper: r.task.promptPaper,
+				ttl:         ttl,
+				ctx:         r.task.ctx,
+			})
+		}
+
+		d := r.fsm.fbo.config.RekeyWithPromptWaitTime()
+		if event.finished.NeedsPaperKey {
+			r.fsm.log.CDebugf(r.task.ctx,
+				"Scheduling rekey due to NeedsPaperKey==true.")
+			return newRekeyStateScheduled(r.fsm, d, rekeyTask{
+				timeout:     &d,
+				promptPaper: true,
+				ttl:         ttl,
+				ctx:         r.task.ctx,
+			})
+		}
+
+		if event.finished.DidRekey {
+			// We enqueue the rekey here again, in case we missed a device due to a
+			// race condition. This is specifically for the situation where user
+			// provisions two devices in a row, and the key update for the 2nd device
+			// only comes in after rekey for a TLF is done, which didn't include the
+			// second device. At this point, there wouldn't be a new MD with rekey
+			// bit set since it's already set. As a result, the TLF won't get rekeyed
+			// for the second device until the next 1-hour timer triggers another
+			// scan.
+			r.fsm.log.CDebugf(r.task.ctx,
+				"Scheduling rekey (recheck) due to DidRekey==true.")
+			return newRekeyStateScheduled(r.fsm, rekeyRecheckInterval, rekeyTask{
+				promptPaper: false,
+				ttl:         ttl,
+				ctx:         r.task.ctx,
+			})
+		}
+
+		r.fsm.log.CDebugf(r.task.ctx,
+			"Not scheduling rekey because no more rekey or recheck is needed.")
+		return newRekeyStateIdle(r.fsm)
+	default:
+		return r
+	}
+}
+
+type listenerForTest struct {
+	repeatedly bool
+	body       func(RekeyEvent)
+}
+
+type rekeyFSM struct {
+	reqs chan RekeyEvent
+	fbo  *folderBranchOps
+	log  logger.Logger
+
+	current rekeyState
+
+	muListenersForTest sync.Mutex
+	listenersForTest   map[rekeyEventType][]listenerForTest
+}
+
+func (m *rekeyFSM) loop() {
+	for e := range m.reqs {
+		if e.eventType == rekeyShutdownEvent {
+			close(m.reqs)
+		}
+
+		next := m.current.reactToEvent(e)
+		m.log.Debug("RekeyFSM transition: %T + %s -> %T",
+			m.current, e, next)
+		m.current = next
+
+		m.triggerCallbacksForTest(e)
+	}
+}
+
+// NewRekeyFSM creates a new rekey FSM.
+func NewRekeyFSM(fbo *folderBranchOps) RekeyFSM {
```

Comment:
Ah, yeah, I guess there are two competing conventions.  In my code I usually put the constructor first no matter what, so it's easy to find.  But I guess it's not fully consistent.